servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/NewRoundRobinLoadBalancer.java

```diff
@@ -250,59 +251,88 @@ public void onNext(@Nullable final Collection<? extends ServiceDiscovererEvent<R
+                        nextHosts.add(host);
+                    } else if (AVAILABLE.equals(event.status())) {
+                        sendReadyEvent = true;
+                        // If the host is already in CLOSED state, we should discard it and create a new entry.
+                        // For duplicate ACTIVE events or for repeated activation due to failed CAS
+                        // of replacing the usedHosts array the marking succeeds so we will not add a new entry.
+                        if (host.markActiveIfNotClosed()) {
+                            nextHosts.add(host);
+                        } else {
+                            nextHosts.add(createHost(event.address()));
+                        }
+                    } else if (EXPIRED.equals(event.status())) {
+                        if (!host.markExpired()) {
+                            nextHosts.add(host);
+                        }
+                    } else if (UNAVAILABLE.equals(event.status())) {
+                        host.markClosed();
+                    } else {
+                        LOGGER.error("{}: Unexpected Status in event:" +
+                                        " {} (mapped to {}). Leaving usedHosts unchanged: {}",
+                                NewRoundRobinLoadBalancer.this, event, event.status(), nextHosts);
+                        nextHosts.add(host);
                     }
-                } else if (usedAddresses.isEmpty()) {
-                    eventStreamProcessor.onNext(LOAD_BALANCER_NOT_READY_EVENT);
+                }
+                // Now process events that didn't have an existing host. The only ones that we actually care
+                // about are the AVAILABLE events which result in a new host.
+                for (ServiceDiscovererEvent<ResolvedAddress> event : eventMap.values()) {
+                    if (AVAILABLE.equals(event.status())) {
+                        sendReadyEvent = true;
+                        nextHosts.add(createHost(event.address()));
+                    }
+                }
+                // We've now built the new list so now we need to CAS it before we can move on. This should only be
+                // racing with closing hosts and closing the whole LB so it shouldn't be common to lose the race.
+                if (usedHostsUpdater.compareAndSet(NewRoundRobinLoadBalancer.this, usedHosts, nextHosts)) {
+                    break;
                 }
             }
 
-            final List<Host<ResolvedAddress, C>> currentHosts = usedHosts;
+            LOGGER.debug("{}: now using addresses (size={}): {}.",
+                    NewRoundRobinLoadBalancer.this, nextHosts.size(), nextHosts);
+            if (nextHosts.isEmpty()) {
+                eventStreamProcessor.onNext(LOAD_BALANCER_NOT_READY_EVENT);
+            } else if (sendReadyEvent) {
+                eventStreamProcessor.onNext(LOAD_BALANCER_READY_EVENT);
```

Comment:
We should not send this event if LB already had hosts before and this event was already sent

Reply:
Good point. On that topic, we don't currently send a NOT_READY event when a host drops out from say the EXPIRED state to UNAVAILABLE state and results in an empty list.

Reply:
That's a good catch! Please file a ticket